src/validation-engine.ts

```diff
@@ -0,0 +1,228 @@
+    return (values: any): Promise<ValidationResult> => {
+      const result = validation(values);
+
+      if (isSyncValidationResult(result)) {
+        return Promise.resolve(result as ValidationResult);
+      } else {
+        return result as Promise<ValidationResult>;
+      }
+    };
+  }
+
+  public isValidationInProgress() {
+    return this.asyncValidationInProgressCount > 0;
+  }
+
+  public validateForm(values: any): Promise<FormValidationResult> {
+    const allValidations = this.fireFieldAndRecordValidations(values);
+    return this.buildValidationResults(allValidations);
+  }
+
+  validateField(
+    values: any,
+    key: string,
+    value: any
+  ): Promise<ValidationResult> {
+    this.asyncValidationInProgressCount++;
+    const asyncValidationPromise = this.fireFieldValidations(
+      values,
+      key,
+      value
+    );
+
+    // TODO: de we need to add some additional check in the promise
+    // e.g. fail or something like that?
+    asyncValidationPromise.then(() => this.asyncValidationInProgressCount--);
+
+    return asyncValidationPromise;
+  }
+
+  private fireFieldAndRecordValidations = (
+    values: any
+  ): Promise<ValidationResult>[] => {
+    let fieldValidationResults: Promise<
+      ValidationResult
+    >[] = fireAllFieldsValidations(
+      values,
+      Object.keys(this.validationsPerField),
+      this.fireFieldValidations
+    );
+
+    // Now global form validations
+    if (this.validationsGlobalForm.length > 0) {
+      fieldValidationResults = [
+        ...fieldValidationResults,
+        ...this.validateGlobalFormValidations(values),
+      ];
+    }
+    return fieldValidationResults;
+  };
+
+  private buildValidationResults = (
+    validations: Promise<ValidationResult>[]
+  ): Promise<FormValidationResult> => {
+    this.asyncValidationInProgressCount++;
+    return new Promise<FormValidationResult>((resolve, reject) => {
+      // Once all the single field validations have been resolved
+      // resolve the fullFormValidatePromise
+      // TODO: Finally issue race condition unit tests
+      Promise.all(validations)
+        .then((fieldValidationResults: ValidationResult[]) => {
+          this.asyncValidationInProgressCount--;
+          const formValidationResult = buildFormValidationResult(
+            fieldValidationResults
+          );
+          resolve(formValidationResult);
+        })
+        .catch(result => {
+          this.asyncValidationInProgressCount--;
+          // Build failed validation Result
+          const errorInformation = `Uncontrolled error when validating full form, check custom validations code`;
+          console.log(errorInformation);
+          reject(result);
+        });
+    });
+  };
+
+  public fireFieldValidations = (
+    values: any,
+    key: string,
+    value: any
+  ): Promise<ValidationResult> => {
+    this.asyncValidationInProgressCount++;
+
+    const fieldValidationResultPromise = new Promise<ValidationResult>(
```

Comment:
Long method maybe we can applt SRP and create some subfunctions?

Reply:
If we remove `validationInProgress` functionality, I think this method looks like:

```
  public fireFieldValidations = (
    values: any,
    key: string,
    value: any
  ): Promise<ValidationResult> =>
    !this.isFieldKeyMappingDefined(key)
      ? Promise.resolve(createDefaultValidationResult())
      : fireSingleFieldValidations(
          values,
          value,
          this.validationsPerField[key]
        ).catch(result => {
          const errorInformation = `Validation Exception in field: ${key}`;
          console.log(errorInformation);
          throw result;
        });
```

Reply:
I would like to keep validationInProgress, React Final form won't use it, but other libraries could take benefit on it, what do you think?

Reply:
Can you elaborate a proposal on this? 

Maybe keep two interface internalValidationResult and ValidationResult?

Do you think it could be a good idea to keep this as a separate case and open a separate branch for this?